src/software/geom/algorithms/intersects.cpp

```diff
@@ -165,3 +165,103 @@ bool intersects(const Segment &first, const Ray &second)
 {
     return intersects(second, first);
 }
+
+bool intersects(const Stadium &first, const Ray &second)
+{
+    auto start = first.segment().getStart();
+    auto end = first.segment().getEnd();
+    auto radius = first.radius();
+
+
+    // The parametric definition of a ray is r(d)=o+tu where o is the origin point and u is a unit vector
+    // Find the distance t along the ray that is closest to start and end
+    // Distance cannot be negative because rays only go positively along the unit vector u
+    auto startDist = std::fmax(0, second.toUnitVector().dot(start - second.getStart()));
+    auto endDist = std::fmax(0, second.toUnitVector().dot(end - second.getStart()));
```

Comment:
nit: I believe `fmax` is mainly for floats and `max` is the more general alternative
```suggestion
    auto startDist = std::max(0, second.toUnitVector().dot(start - second.getStart()));
    auto endDist = std::max(0, second.toUnitVector().dot(end - second.getStart()));
```